snuba/state/rate_limit.py

```diff
@@ -0,0 +1,181 @@
+from collections import namedtuple
+from contextlib import AbstractContextManager, ExitStack
+from dataclasses import dataclass
+import logging
+import time
+from types import TracebackType
+from typing import Mapping, MutableMapping, Optional, Sequence, Type
+import uuid
+
+from snuba import state
+
+logger = logging.getLogger('snuba.state.rate_limit')
+
+
+@dataclass(frozen=True)
+class RateLimitParameters:
+    """
+    The configuration object which defines all the needed properties to create
+    a rate limit.
+    """
+
+    rate_limit_name: str
+    bucket: str
+    per_second_limit: Optional[float]
+    concurrent_limit: Optional[float]
+
+
+class RateLimitExceeded(Exception):
+    """
+    Exception thrown when the rate limit is exceeded
+    """
+
+
+class RateLimit(AbstractContextManager):
```

Comment:
Since you do not have subclasses of this one anymore, do you still need to create a class for the context manager? I think you can move back to the context manager method in the previous implementation which is simpler and reduces the scope of the PR as well.

Reply:
Makes sense.

Do you think I should remove the raising, trying and catching too? I feel that's an improvement to the previous implementation cause that's how context managers should behave. So, thinking I should leave that in there.

Reply:
Yes, I would preserve the try/catch.